src/main/java/io/cryostat/recordings/JvmIdHelper.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright The Cryostat Authors
+ *
+ * The Universal Permissive License (UPL), Version 1.0
+ *
+ * Subject to the condition set forth below, permission is hereby granted to any
+ * person obtaining a copy of this software, associated documentation and/or data
+ * (collectively the "Software"), free of charge and under any and all copyright
+ * rights in the Software, and any and all patent rights owned or freely
+ * licensable by each licensor hereunder covering either (i) the unmodified
+ * Software as contributed to or provided by such licensor, or (ii) the Larger
+ * Works (as defined below), to deal in both
+ *
+ * (a) the Software, and
+ * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+ * one is included with the Software (each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.recordings;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import io.cryostat.configuration.CredentialsManager;
+import io.cryostat.core.log.Logger;
+import io.cryostat.net.ConnectionDescriptor;
+import io.cryostat.net.TargetConnectionManager;
+
+public class JvmIdHelper {
+    private final TargetConnectionManager targetConnectionManager;
+    private final CredentialsManager credentialsManager;
+    private final Logger logger;
+
+    private final Map<String, String> jvmIdMap;
+
+    public static final int CONNECT_TIMEOUT_SECONDS = 1;
```

Comment:
I think this timeout should be longer by default, and it should probably be configurable.

Maybe just [reuse this](https://github.com/cryostatio/cryostat/blob/b688e76759fd94a87340206ed7785e777f552757/src/main/java/io/cryostat/net/web/http/HttpModule.java#L65)?

Reply:
For some reason which I don't know why, connecting to targets is really flaky, so often requests will timeout - even if the timeout is set to the default generation timeout seconds = 29s. Going to the web-client and looking at recordings of targets will hang for this amount of time. I sort of have a feeling its because lots of different components are now calling to compute the jvmId during startup and there's some sort of deadlock, maybe with the `TargetConnectionManager` locks or something else,  but I can't really pinpoint why. If I set the timeout to something like 1 second, some of the first few requests may get an early timeout but eventually for some reason all the targets and cryostat behave normally. Any insight?

Reply:
It definitely could be related to a deadlock or other threading issue. Can you reproduce this issue on `main`? Do you have any custom target definitions or just the standard samples from `smoketest.sh`?

Reply:
I only use everything from smoketest.sh, I will try recreating from main.

Reply:
In `main`
```bash
$ mvn clean package
$ CRYOSTAT_DISABLE_SSL=true sh smoketest.sh > conf/t.log # to see log easily
```
Sometimes it happens, sometimes it doesn't. This example is one instance that does:

```
Sep 20, 2022 11:01:02 PM io.cryostat.core.log.Logger info
INFO: Outgoing WS message: {"meta":{"category":"TargetJvmDiscovery","type":{"type":"application","subType":"json"},"serverTime":1663714862},"message":{"event":{"serviceRef":{"connectUrl":"service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi","alias":"es.andrewazor.demo.Main","labels":{},"annotations":{"platform":{},"cryostat":{"HOST":"cryostat","PORT":"9093","JAVA_MAIN":"es.andrewazor.demo.Main","REALM":"JDP"}}},"kind":"FOUND"}}}
Sep 20, 2022 11:01:02 PM io.cryostat.core.log.Logger info
INFO: Target service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi did not have a jvmId
Sep 20, 2022 11:01:02 PM io.cryostat.core.log.Logger info
INFO: Creating connection for service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi
Hibernate: 
    /* select
        generatedAlias0 
    from
        PluginInfo as generatedAlias0 */ select
            plugininfo0_.id as id1_0_,
            plugininfo0_.callback as callback2_0_,
            plugininfo0_.realm as realm3_0_,
            plugininfo0_.subtree as subtree4_0_ 
        from
            PluginInfo plugininfo0_
Sep 20, 2022 11:01:03 PM io.cryostat.core.log.Logger info
INFO: Creating connection for service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi
Sep 20, 2022 11:01:12 PM io.cryostat.core.log.Logger info
INFO: Removing cached connection for service:jmx:rmi:///jndi/rmi://cryostat:10000/jmxrmi: EXPIRED
Sep 20, 2022 11:01:12 PM io.cryostat.core.log.Logger info
INFO: Connection for service:jmx:rmi:///jndi/rmi://cryostat:10000/jmxrmi closed
Sep 20, 2022 11:01:15 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:58508): POST /api/v2.1/auth 200 9ms
Sep 20, 2022 11:01:15 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:58516): GET /api/v1/notifications_url 200 20ms
Sep 20, 2022 11:01:15 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:58528): GET /health 200 287ms
Sep 20, 2022 11:01:16 PM io.cryostat.core.log.Logger info
INFO: Connected remote client 10.0.2.100:58534
Sep 20, 2022 11:01:16 PM io.cryostat.core.log.Logger info
INFO: Outgoing WS message: {"meta":{"category":"WsClientActivity","type":{"type":"application","subType":"json"},"serverTime":1663714876},"message":{"10.0.2.100:58534":"connected"}}
Sep 20, 2022 11:01:16 PM io.cryostat.core.log.Logger info
INFO: Authenticated remote client 10.0.2.100:58534
Sep 20, 2022 11:01:16 PM io.cryostat.core.log.Logger info
INFO: Outgoing WS message: {"meta":{"category":"WsClientActivity","type":{"type":"application","subType":"json"},"serverTime":1663714876},"message":{"10.0.2.100:58534":"accepted"}}
Sep 20, 2022 11:01:16 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:58508): GET /api/v1/grafana_datasource_url 200 19ms
Sep 20, 2022 11:01:16 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:58528): GET /api/v1/grafana_dashboard_url 200 33ms
Sep 20, 2022 11:01:16 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:58508): GET /api/v1/recordings 200 20ms
Hibernate: 
    /* select
        generatedAlias0 
    from
        PluginInfo as generatedAlias0 */ select
            plugininfo0_.id as id1_0_,
            plugininfo0_.callback as callback2_0_,
            plugininfo0_.realm as realm3_0_,
            plugininfo0_.subtree as subtree4_0_ 
        from
            PluginInfo plugininfo0_
Sep 20, 2022 11:01:16 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:58508): GET /api/v1/targets 200 18ms
Hibernate: 
    /* select
        generatedAlias0 
    from
        PluginInfo as generatedAlias0 */ select
            plugininfo0_.id as id1_0_,
            plugininfo0_.callback as callback2_0_,
            plugininfo0_.realm as realm3_0_,
            plugininfo0_.subtree as subtree4_0_ 
        from
            PluginInfo plugininfo0_
Sep 20, 2022 11:01:16 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:58508): POST /api/v2.2/graphql 200 75ms
Sep 20, 2022 11:02:01 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:39686): POST /api/v2.2/discovery 200 30027ms
Sep 20, 2022 11:02:01 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:39698): POST /api/v2.2/discovery 200 30028ms
Sep 20, 2022 11:02:31 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:37028): POST /api/v2.2/discovery 200 30028ms
Sep 20, 2022 11:02:31 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:37034): POST /api/v2.2/discovery 200 30028ms
Sep 20, 2022 11:03:01 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:57954): POST /api/v2.2/discovery 200 30028ms
Sep 20, 2022 11:03:01 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:57956): POST /api/v2.2/discovery 200 30028ms
Sep 20, 2022 11:03:31 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:60210): POST /api/v2.2/discovery 200 30028ms
Sep 20, 2022 11:03:31 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:60222): POST /api/v2.2/discovery 200 30028ms
```
Since  there's no timeout in main, it will hang forever. And we can see that the connection 
`INFO: Creating connection for service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi` happens in the `RecordingMetadataManager` listener accepts for computing the jvmId, and the connection never closes.

Reply:
Interesting. Could it be that the connection failed and the exception just wasn't logged? But I guess in that case, we should see the `TargetConnectionManager` logging the connection being closed and evicted from cache.

Stepping through the `JFRConnection`, it definitely looks like it's possible for the JMX connection to silently hang:

https://github.com/cryostatio/cryostat-core/blob/65f662837fb0eae0fcb9cf8b3a3ed7027ca53cb9/src/main/java/io/cryostat/core/net/JFRConnection.java#L295

and I don't see any nice way to apply a timeout here with the underlying JMC RJMXConnection code. Hmm... :-/

Reply:
This also seems noteworthy:

```
INFO: Target service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi did not have a jvmId
Sep 20, 2022 11:01:02 PM io.cryostat.core.log.Logger info
INFO: Creating connection for service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi
Hibernate: 
    /* select
        generatedAlias0 
    from
        PluginInfo as generatedAlias0 */ select
            plugininfo0_.id as id1_0_,
            plugininfo0_.callback as callback2_0_,
            plugininfo0_.realm as realm3_0_,
            plugininfo0_.subtree as subtree4_0_ 
        from
            PluginInfo plugininfo0_
Sep 20, 2022 11:01:03 PM io.cryostat.core.log.Logger info
INFO: Creating connection for service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi
```

We're trying to open two connections to the same target application in very close succession, and then we never see either of those connections close. The `TargetConnectionManager` only logs that line when it's actually trying to open a new JMX connection, not just when something tries to perform a task requiring a connection, since it does try to cache and reuse connections. So this definitely could be some concurrency issue, though I don't yet fully understand why/how.

Reply:
I guess this could be originating from the `TargetConnectionManager.executeConnectedTask(_, _, *false*)` calls which intentionally skip using the cache and open a new connection all the time. I know I wrote the automated rules processor to use that, and I think the metadata processing is using it too. The reason for doing that with the automated rules was just to avoid filling the cache with connections that may not get reused and wasting memory on them, or worse filling the cache and causing premature evictions in the case where the number of total targets and the rate at which they can be processed exceeds the maximum size of the cache.

I guess the next thing to try is to ensure that the connection cache has no size limit and to force every connection attempt to go through the cache, even the ones for automated rules and metadata pruning. This should funnel everything through the cache and locking logic and ensure we only try to open one JMX connection per target at a time. I don't know why opening two connections would cause this failure, though. I also wonder if we need to apply some kind of rate limiting to our outbound connections so we aren't trying to open *N* connections for each of *N* targets all at once at startup.

Reply:
Hmm... I'll try these ideas out, thanks a lot! I wonder why having two connections at once would result in this as well, seems like it shouldn't happen.

Reply:
I finally figured it out, I was just violating vertx's golden rule - blocking the event loop by calling CompletableFuture.get() when trying to compute the jvmId. So in actuality, the ids were getting computed, but by blocking the main thread, I was also blocking everything else, including actually computing the ids so it would always timeout. So to fix this, all I had to do was just remove the completable futures and other new thread making procedures like vertx.executeBlocking. :\

Reply:
That makes sense, though `vertx.executeBlocking` is not actually a blocking call on its own. It just submits a unit of work to the verticle's worker pool for execution. You do need to be mindful of `Future`s and `Promise`s to make sure you're actually treating them asynchronously though, and not forking a task off to a separate thread and on the original thread immediately calling `.get()` to retrieve the result. I know we're guilty of that in a few places right now - I think the `RecordingHelper` classes do this, for example. I haven't had the time to go through and refactor those to be properly async.

Reply:
related: https://github.com/cryostatio/cryostat/pull/1080